tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc

```diff
@@ -276,17 +274,34 @@ struct HloToLhloReduceOpConverter : public BaseOpConversion<mhlo::ReduceOp> {
     // Convert the region signature to memref and add extra result.
     auto& entry_block = new_op.body().front();
     TypeConverter::SignatureConversion sig_conversion(
-        entry_block.getNumArguments() + 1);
+        entry_block.getNumArguments() / 2 * 3);
     for (auto arg : entry_block.getArguments()) {
       auto old_type = arg.getType().cast<TensorType>();
       auto new_type =
           MemRefType::get(old_type.getShape(), old_type.getElementType());
       sig_conversion.addInputs(arg.getArgNumber(), new_type);
     }
     auto return_op = cast<mhlo::ReturnOp>(entry_block.getTerminator());
-    auto result_type = return_op.results().front().getType().cast<TensorType>();
-    sig_conversion.addInputs({MemRefType::get(result_type.getShape(),
-                                              result_type.getElementType())});
+    if (auto tuple_ty =
+            return_op.results().front().getType().dyn_cast<TupleType>()) {
+      auto tuple_op = return_op.getODSOperands(0).front().getDefiningOp();
+      return_op.getOperation()->dropAllReferences();
+      rewriter.eraseOp(tuple_op);
```

Comment:
Do we have to delete the tuple op manually? I thought it will get dropped if it is unused.

Reply:
The tuple_op is already on the rewriter's worklist. If I don't delete it manually, I will get ```failed to legalize operation 'mhlo.tuple' that was explicitly marked illegal```